perf(semantic): mark `checker::check` as `inline(always)`

camc314 opened the pull request:
the reasoning here is:
1. `check` is a big function with a big match arm
2. if we inline it into leave_node
3. and inline leave_node into it's caller
4. the match will be flattened, as the compiler knows what varient of the kind it is, this means all the other match arms can be dropped, as well as the match it's self, just leaving the call into the check fn (if needed) <- if this works correctly, it also means this change will hardly have any impact on binary size

Boshen reviewed (commented):
Add two inline comments to these two `#[inline(always)]` and we should be good to go!

camc314 committed:
perf(semantic): mark `checker::check` as `inline(always)` (#17459)

the reasoning here is:
1. `check` is a big function with a big match arm
2. if we inline it into leave_node
3. and inline leave_node into it's caller
4. the match will be flattened, as the compiler knows what varient of the kind it is, this means all the other match arms can be dropped, as well as the match it's self, just leaving the call into the check fn (if needed) <- if this works correctly, it also means this change will hardly have any impact on binary size
